common/base/include/claragenomics/utils/threadsafe_containers.hpp

```diff
@@ -0,0 +1,222 @@
+
+    /// \brief deleted copy assignment operator
+    /// \param rhs
+    ThreadsafeProducerConsumer& operator==(const ThreadsafeProducerConsumer& rhs) = delete;
+
+    /// \brief deleted move constructor
+    /// \param rhs
+    ThreadsafeProducerConsumer(ThreadsafeProducerConsumer&& rhs) = delete; // could be implemeted if needed
+
+    /// \brief deleted move assignment operator
+    /// \param rhs
+    ThreadsafeProducerConsumer& operator==(ThreadsafeProducerConsumer&& rhs) = delete; // could be implemeted if needed
+
+    /// \brief destructor
+    ~ThreadsafeProducerConsumer() = default;
+
+    /// \brief adds an element to the queue
+    ///
+    /// \param element element to add
+    /// \throw std::logic_error if called after signal_pushed_last_element() has been called by any producer
+    void add_new_element(const T& element) // not using reference in order to support both rvalue and lvalue
+    {
+        // consider adding max number of elements and waiting for consumer to process an element before adding a new one
+        std::lock_guard<std::mutex> lg(mutex_);
+        if (pushed_last_element_)
+        {
+            throw std::logic_error("ThreadsafeProducerConsumer: pushed an element after signal_pushed_last_element() has been called");
+        }
+        data_.push_front(element);
+        condition_variable_.notify_all();
+    }
+
+    /// \brief adds an element to the queue
+    ///
+    /// \param element element to add
+    /// \throw std::logic_error if called after signal_pushed_last_element() has been called by any producer
+    void add_new_element(T&& element) // not using reference in order to support both rvalue and lvalue
+    {
+        // consider adding max number of elements and waiting for consumer to process an element before adding a new one
+        std::lock_guard<std::mutex> lg(mutex_);
+        if (pushed_last_element_)
+        {
+            throw std::logic_error("ThreadsafeProducerConsumer: pushed an element after signal_pushed_last_element() has been called");
+        }
+        data_.push_front(std::move(element));
+        condition_variable_.notify_all();
+    }
+
+    /// \brief tells container that no new elements are going to be added
+    ///
+    /// After this method has been called and all elements have been consumed get_next_element() will be returning
+    /// empty optionals to indicate that there will be no new elements and that consumers can finish processing
+    ///
+    /// \throw std::logic_error if called after any producer has already called this method
+    void signal_pushed_last_element()
+    {
+        std::lock_guard<std::mutex> lg(mutex_);
+        if (pushed_last_element_)
+        {
+            throw std::logic_error("ThreadsafeProducerConsumer: called signal_pushed_last_element() more than once");
+        }
+        pushed_last_element_ = true;
+        condition_variable_.notify_all();
```

Comment:
In this case `notify_all()` is needed. Imagine several consumers waiting on an empty queue and producer calling `signal_pushed_last_element()`. In that case we would like to notify all consumers that there will be no new data coming in, otherwise some of them would keep waiting

Reply:
Done: moved `notify_all()` it outside of mutex

Reply:
Ah yes, right... But at least move the `notify` outside still applied. :phew: :)